File: src/lib/mpris.zig
const dbus = @import("dbus");
const std = @import("std");
const Allocator = std.mem.Allocator;

pub const PlaybackStatus = enum {
    Playing,
    Paused,
    Stopped,

    pub fn fromString(str: []const u8) ?PlaybackStatus {
        if (std.mem.eql(u8, str, "Playing")) return .Playing;
        if (std.mem.eql(u8, str, "Paused")) return .Paused;
        if (std.mem.eql(u8, str, "Stopped")) return .Stopped;
        return null;
    }

    pub fn toString(self: PlaybackStatus) []const u8 {
        return switch (self) {
            .Playing => "Playing",
            .Paused => "Paused",
            .Stopped => "Stopped",
        };
    }
};

pub const LoopStatus = enum {
    None,
    Track,
    Playlist,

    pub fn fromString(str: []const u8) ?LoopStatus {
        if (std.mem.eql(u8, str, "None")) return .None;
        if (std.mem.eql(u8, str, "Track")) return .Track;
        if (std.mem.eql(u8, str, "Playlist")) return .Playlist;
        return null;
    }

    pub fn toString(self: LoopStatus) []const u8 {
        return switch (self) {
            .None => "None",
            .Track => "Track",
            .Playlist => "Playlist",
        };
    }
};

pub const Metadata = struct {
    trackid: ?[]const u8 = null,
    length: ?i64 = null,
    artUrl: ?[]const u8 = null,
    album: ?[]const u8 = null,
    albumArtist: ?[][]const u8 = null,
    artist: ?[][]const u8 = null,
    asText: ?[]const u8 = null,
    audioBPM: ?i32 = null,
    autoRating: ?f64 = null,
    comment: ?[][]const u8 = null,
    composer: ?[][]const u8 = null,
    contentCreated: ?[]const u8 = null,
    discNumber: ?i32 = null,
    firstUsed: ?[]const u8 = null,
    genre: ?[][]const u8 = null,
    lastUsed: ?[]const u8 = null,
    lyricist: ?[][]const u8 = null,
    title: ?[]const u8 = null,
    trackNumber: ?i32 = null,
    url: ?[]const u8 = null,
    useCount: ?i32 = null,
    userRating: ?f64 = null,

    pub fn deinit(self: Metadata, allocator: Allocator) void {
        if (self.trackid) |v| allocator.free(v);
        if (self.artUrl) |v| allocator.free(v);
        if (self.album) |v| allocator.free(v);
        if (self.albumArtist) |arr| {
            for (arr) |v| allocator.free(v);
            allocator.free(arr);
        }
        if (self.artist) |arr| {
            for (arr) |v| allocator.free(v);
            allocator.free(arr);
        }
        if (self.asText) |v| allocator.free(v);
        if (self.comment) |arr| {
            for (arr) |v| allocator.free(v);
            allocator.free(arr);
        }
        if (self.composer) |arr| {
            for (arr) |v| allocator.free(v);
            allocator.free(arr);
        }
        if (self.contentCreated) |v| allocator.free(v);
        if (self.firstUsed) |v| allocator.free(v);
        if (self.genre) |arr| {
            for (arr) |v| allocator.free(v);
            allocator.free(arr);
        }
        if (self.lastUsed) |v| allocator.free(v);
        if (self.lyricist) |arr| {
            for (arr) |v| allocator.free(v);
            allocator.free(arr);
        }
        if (self.title) |v| allocator.free(v);
        if (self.url) |v| allocator.free(v);
    }
};

pub const PlayerInfo = struct {
    busName: []const u8,
    identity: ?[]const u8 = null,
    desktopEntry: ?[]const u8 = null,
    canQuit: bool = false,
    canRaise: bool = false,
    canSetFullscreen: bool = false,
    hasTrackList: bool = false,
    fullscreen: bool = false,

    pub fn deinit(self: PlayerInfo, allocator: Allocator) void {
        allocator.free(self.busName);
        if (self.identity) |v| allocator.free(v);
        if (self.desktopEntry) |v| allocator.free(v);
    }
};

pub const PlayerStatus = struct {
    canControl: bool = false,
    canGoNext: bool = false,
    canGoPrevious: bool = false,
    canPause: bool = false,
    canPlay: bool = false,
    canSeek: bool = false,
    loopStatus: LoopStatus = .None,
    maximumRate: f64 = 1.0,
    metadata: ?Metadata = null,
    minimumRate: f64 = 1.0,
    playbackStatus: PlaybackStatus = .Stopped,
    position: i64 = 0,
    rate: f64 = 1.0,
    shuffle: bool = false,
    volume: f64 = 1.0,

    pub fn deinit(self: PlayerStatus, allocator: Allocator) void {
        if (self.metadata) |m| m.deinit(allocator);
    }
};

pub const MediaPlayer2 = struct {
    const Self = @This();
    allocator: Allocator,
    bus: *dbus.Bus,
    busName: []const u8,
    player: *dbus.Object,
    playerInterface: *dbus.Object,
    userdata: ?*anyopaque = null,

    pub fn init(allocator: Allocator, bus: *dbus.Bus, busName: []const u8) !Self {
        const name = try allocator.dupe(u8, busName);
        errdefer allocator.free(name);

        const player = try bus.proxy(name, "/org/mpris/MediaPlayer2", "org.mpris.MediaPlayer2");
        errdefer player.deinit();

        const playerInterface = try bus.proxy(name, "/org/mpris/MediaPlayer2", "org.mpris.MediaPlayer2.Player");
        errdefer playerInterface.deinit();

        return .{
            .allocator = allocator,
            .bus = bus,
            .busName = name,
            .player = player,
            .playerInterface = playerInterface,
        };
    }

    pub fn deinit(self: Self) void {
        self.player.deinit();
        self.playerInterface.deinit();
        self.allocator.free(self.busName);
    }

    // org.mpris.MediaPlayer2 methods
    pub fn raise(self: Self) !void {
        const result = try self.player.call("Raise", .{}, .{});
        defer result.deinit();
    }

    pub fn quit(self: Self) !void {
        const result = try self.player.call("Quit", .{}, .{});
        defer result.deinit();
    }

    // org.mpris.MediaPlayer2.Player methods
    pub fn next(self: Self) !void {
        const result = try self.playerInterface.call("Next", .{}, .{});
        defer result.deinit();
    }

    pub fn previous(self: Self) !void {
        const result = try self.playerInterface.call("Previous", .{}, .{});
        defer result.deinit();
    }

    pub fn pause(self: Self) !void {
        const result = try self.playerInterface.call("Pause", .{}, .{});
        defer result.deinit();
    }

    pub fn playPause(self: Self) !void {
        const result = try self.playerInterface.call("PlayPause", .{}, .{});
        defer result.deinit();
    }

    pub fn stop(self: Self) !void {
        const result = try self.playerInterface.call("Stop", .{}, .{});
        defer result.deinit();
    }

    pub fn play(self: Self) !void {
        const result = try self.playerInterface.call("Play", .{}, .{});
        defer result.deinit();
    }

    pub fn seek(self: Self, offset: i64) !void {
        const result = try self.playerInterface.call("Seek", .{dbus.Int64}, .{offset});
        defer result.deinit();
    }

    pub fn setPosition(self: Self, trackId: []const u8, position: i64) !void {
        const result = try self.playerInterface.call("SetPosition", .{ dbus.ObjectPath, dbus.Int64 }, .{ trackId, position });
        defer result.deinit();
    }

    pub fn openUri(self: Self, uri: []const u8) !void {
        const result = try self.playerInterface.call("OpenUri", .{dbus.String}, .{uri});
        defer result.deinit();
    }

    // Property getters for org.mpris.MediaPlayer2
    pub fn getIdentity(self: Self, allocator: Allocator) ![]const u8 {
        return try self.player.getAlloc(allocator, "Identity", dbus.String);
    }

    pub fn getDesktopEntry(self: Self, allocator: Allocator) ![]const u8 {
        return try self.player.getAlloc(allocator, "DesktopEntry", dbus.String);
    }

    pub fn canQuit(self: Self) !bool {
        return try self.player.getBasic("CanQuit", dbus.Boolean);
    }

    pub fn canRaise(self: Self) !bool {
        return try self.player.getBasic("CanRaise", dbus.Boolean);
    }

    pub fn canSetFullscreen(self: Self) !bool {
        return try self.player.getBasic("CanSetFullscreen", dbus.Boolean);
    }

    pub fn hasTrackList(self: Self) !bool {
        return try self.player.getBasic("HasTrackList", dbus.Boolean);
    }

    pub fn getFullscreen(self: Self) !bool {
        return try self.player.getBasic("Fullscreen", dbus.Boolean);
    }

    pub fn setFullscreen(self: Self, fullscreen: bool) !void {
        try self.player.set("Fullscreen", dbus.Boolean, fullscreen);
    }

    // Property getters for org.mpris.MediaPlayer2.Player
    pub fn getPlaybackStatus(self: Self, allocator: Allocator) !PlaybackStatus {
        const str = try self.playerInterface.getAlloc(allocator, "PlaybackStatus", dbus.String);
        defer allocator.free(str);
        return PlaybackStatus.fromString(str) orelse .Stopped;
    }

    pub fn getLoopStatus(self: Self, allocator: Allocator) !LoopStatus {
        const str = try self.playerInterface.getAlloc(allocator, "LoopStatus", dbus.String);
        defer allocator.free(str);
        return LoopStatus.fromString(str) orelse .None;
    }

    pub fn setLoopStatus(self: Self, status: LoopStatus) !void {
        try self.playerInterface.set("LoopStatus", dbus.String, status.toString());
    }

    pub fn getRate(self: Self) !f64 {
        return try self.playerInterface.getBasic("Rate", dbus.Double);
    }

    pub fn setRate(self: Self, rate: f64) !void {
        try self.playerInterface.set("Rate", dbus.Double, rate);
    }

    pub fn getShuffle(self: Self) !bool {
        return try self.playerInterface.getBasic("Shuffle", dbus.Boolean);
    }

    pub fn setShuffle(self: Self, shuffle: bool) !void {
        try self.playerInterface.set("Shuffle", dbus.Boolean, shuffle);
    }

    pub fn getVolume(self: Self) !f64 {
        return try self.playerInterface.getBasic("Volume", dbus.Double);
    }

    pub fn setVolume(self: Self, volume: f64) !void {
        try self.playerInterface.set("Volume", dbus.Double, volume);
    }

    pub fn getPosition(self: Self) !i64 {
        return try self.playerInterface.getBasic("Position", dbus.Int64);
    }

    pub fn getMinimumRate(self: Self) !f64 {
        return try self.playerInterface.getBasic("MinimumRate", dbus.Double);
    }

    pub fn getMaximumRate(self: Self) !f64 {
        return try self.playerInterface.getBasic("MaximumRate", dbus.Double);
    }

    pub fn canGoNext(self: Self) !bool {
        return try self.playerInterface.getBasic("CanGoNext", dbus.Boolean);
    }

    pub fn canGoPrevious(self: Self) !bool {
        return try self.playerInterface.getBasic("CanGoPrevious", dbus.Boolean);
    }

    pub fn canPlay(self: Self) !bool {
        return try self.playerInterface.getBasic("CanPlay", dbus.Boolean);
    }

    pub fn canPause(self: Self) !bool {
        return try self.playerInterface.getBasic("CanPause", dbus.Boolean);
    }

    pub fn canSeek(self: Self) !bool {
        return try self.playerInterface.getBasic("CanSeek", dbus.Boolean);
    }

    pub fn canControl(self: Self) !bool {
        return try self.playerInterface.getBasic("CanControl", dbus.Boolean);
    }

    pub fn getMetadata(self: Self, allocator: Allocator) !Metadata {
        const result = try self.playerInterface.get("Metadata", dbus.Vardict);
        defer result.deinit();

        var metadata = Metadata{};
        errdefer metadata.deinit(allocator);

        for (result.value) |entry| {
            const key = entry.key;
            const value = entry.value;

            if (std.mem.eql(u8, key, "mpris:trackid")) {
                if (value.tag == dbus.String.tag) {
                    metadata.trackid = try allocator.dupe(u8, value.as(dbus.String));
                }
            } else if (std.mem.eql(u8, key, "mpris:length")) {
                if (value.tag == dbus.Int64.tag) {
                    metadata.length = value.as(dbus.Int64);
                }
            } else if (std.mem.eql(u8, key, "mpris:artUrl")) {
                if (value.tag == dbus.String.tag) {
                    metadata.artUrl = try allocator.dupe(u8, value.as(dbus.String));
                }
            } else if (std.mem.eql(u8, key, "xesam:album")) {
                if (value.tag == dbus.String.tag) {
                    metadata.album = try allocator.dupe(u8, value.as(dbus.String));
                }
            } else if (std.mem.eql(u8, key, "xesam:albumArtist")) {
                // Handle array of strings
                if (value.tag == dbus.Array(dbus.String).tag) {
                    const artists = value.as(dbus.Array(dbus.String));
                    if (artists.len > 0) {
                        var duped = try allocator.alloc([]const u8, artists.len);
                        errdefer allocator.free(duped);
                        for (artists, 0..) |artist, i| {
                            duped[i] = try allocator.dupe(u8, artist);
                        }
                        metadata.albumArtist = duped;
                    }
                }
            } else if (std.mem.eql(u8, key, "xesam:artist")) {
                // Handle array of strings
                if (value.tag == dbus.Array(dbus.String).tag) {
                    const artists = value.as(dbus.Array(dbus.String));
                    if (artists.len > 0) {
                        var duped = try allocator.alloc([]const u8, artists.len);
                        errdefer allocator.free(duped);
                        for (artists, 0..) |artist, i| {
                            duped[i] = try allocator.dupe(u8, artist);
                        }
                        metadata.artist = duped;
                    }
                }
            } else if (std.mem.eql(u8, key, "xesam:asText")) {
                if (value.tag == dbus.String.tag) {
                    metadata.asText = try allocator.dupe(u8, value.as(dbus.String));
                }
            } else if (std.mem.eql(u8, key, "xesam:audioBPM")) {
                if (value.tag == dbus.Int32.tag) {
                    metadata.audioBPM = value.as(dbus.Int32);
                }
            } else if (std.mem.eql(u8, key, "xesam:autoRating")) {
                if (value.tag == dbus.Double.tag) {
                    metadata.autoRating = value.as(dbus.Double);
                }
            } else if (std.mem.eql(u8, key, "xesam:comment")) {
                if (value.tag == dbus.Array(dbus.String).tag) {
                    const comments = value.as(dbus.Array(dbus.String));
                    if (comments.len > 0) {
                        var duped = try allocator.alloc([]const u8, comments.len);
                        errdefer allocator.free(duped);
                        for (comments, 0..) |comment, i| {
                            duped[i] = try allocator.dupe(u8, comment);
                        }
                        metadata.comment = duped;
                    }
                }
            } else if (std.mem.eql(u8, key, "xesam:composer")) {
                if (value.tag == dbus.Array(dbus.String).tag) {
                    const composers = value.as(dbus.Array(dbus.String));
                    if (composers.len > 0) {
                        var duped = try allocator.alloc([]const u8, composers.len);
                        errdefer allocator.free(duped);
                        for (composers, 0..) |composer, i| {
                            duped[i] = try allocator.dupe(u8, composer);
                        }
                        metadata.composer = duped;
                    }
                }
            } else if (std.mem.eql(u8, key, "xesam:contentCreated")) {
                if (value.tag == dbus.String.tag) {
                    metadata.contentCreated = try allocator.dupe(u8, value.as(dbus.String));
                }
            } else if (std.mem.eql(u8, key, "xesam:discNumber")) {
                if (value.tag == dbus.Int32.tag) {
                    metadata.discNumber = value.as(dbus.Int32);
                }
            } else if (std.mem.eql(u8, key, "xesam:firstUsed")) {
                if (value.tag == dbus.String.tag) {
                    metadata.firstUsed = try allocator.dupe(u8, value.as(dbus.String));
                }
            } else if (std.mem.eql(u8, key, "xesam:genre")) {
                if (value.tag == dbus.Array(dbus.String).tag) {
                    const genres = value.as(dbus.Array(dbus.String));
                    if (genres.len > 0) {
                        var duped = try allocator.alloc([]const u8, genres.len);
                        errdefer allocator.free(duped);
                        for (genres, 0..) |genre, i| {
                            duped[i] = try allocator.dupe(u8, genre);
                        }
                        metadata.genre = duped;
                    }
                }
            } else if (std.mem.eql(u8, key, "xesam:lastUsed")) {
                if (value.tag == dbus.String.tag) {
                    metadata.lastUsed = try allocator.dupe(u8, value.as(dbus.String));
                }
            } else if (std.mem.eql(u8, key, "xesam:lyricist")) {
                if (value.tag == dbus.Array(dbus.String).tag) {
                    const lyricists = value.as(dbus.Array(dbus.String));
                    if (lyricists.len > 0) {
                        var duped = try allocator.alloc([]const u8, lyricists.len);
                        errdefer allocator.free(duped);
                        for (lyricists, 0..) |lyricist, i| {
                            duped[i] = try allocator.dupe(u8, lyricist);
                        }
                        metadata.lyricist = duped;
                    }
                }
            } else if (std.mem.eql(u8, key, "xesam:title")) {
                if (value.tag == dbus.String.tag) {
                    metadata.title = try allocator.dupe(u8, value.as(dbus.String));
                }
            } else if (std.mem.eql(u8, key, "xesam:trackNumber")) {
                if (value.tag == dbus.Int32.tag) {
                    metadata.trackNumber = value.as(dbus.Int32);
                }
            } else if (std.mem.eql(u8, key, "xesam:url")) {
                if (value.tag == dbus.String.tag) {
                    metadata.url = try allocator.dupe(u8, value.as(dbus.String));
                }
            } else if (std.mem.eql(u8, key, "xesam:useCount")) {
                if (value.tag == dbus.Int32.tag) {
                    metadata.useCount = value.as(dbus.Int32);
                }
            } else if (std.mem.eql(u8, key, "xesam:userRating")) {
                if (value.tag == dbus.Double.tag) {
                    metadata.userRating = value.as(dbus.Double);
                }
            }
        }

        return metadata;
    }

    pub fn getPlayerInfo(self: Self, allocator: Allocator) !PlayerInfo {
        var info = PlayerInfo{
            .busName = try allocator.dupe(u8, self.busName),
        };
        errdefer info.deinit(allocator);

        info.identity = self.getIdentity(allocator) catch null;
        info.desktopEntry = self.getDesktopEntry(allocator) catch null;
        info.canQuit = self.canQuit() catch false;
        info.canRaise = self.canRaise() catch false;
        info.canSetFullscreen = self.canSetFullscreen() catch false;
        info.hasTrackList = self.hasTrackList() catch false;
        info.fullscreen = self.getFullscreen() catch false;

        return info;
    }

    pub fn getPlayerStatus(self: Self, allocator: Allocator) !PlayerStatus {
        var status = PlayerStatus{};
        errdefer status.deinit(allocator);

        status.playbackStatus = self.getPlaybackStatus(allocator) catch .Stopped;
        status.loopStatus = self.getLoopStatus(allocator) catch .None;
        status.rate = self.getRate() catch 1.0;
        status.shuffle = self.getShuffle() catch false;
        status.volume = self.getVolume() catch 1.0;
        status.position = self.getPosition() catch 0;
        status.minimumRate = self.getMinimumRate() catch 1.0;
        status.maximumRate = self.getMaximumRate() catch 1.0;
        status.canGoNext = self.canGoNext() catch false;
        status.canGoPrevious = self.canGoPrevious() catch false;
        status.canPlay = self.canPlay() catch false;
        status.canPause = self.canPause() catch false;
        status.canSeek = self.canSeek() catch false;
        status.canControl = self.canControl() catch false;
        status.metadata = self.getMetadata(allocator) catch null;

        return status;
    }
};

pub fn listPlayers(allocator: Allocator, bus: *dbus.Bus) ![][]const u8 {
    const dbusDaemon = try bus.proxy("org.freedesktop.DBus", "/org/freedesktop/DBus", "org.freedesktop.DBus");
    defer dbusDaemon.deinit();

    const result = try dbusDaemon.call("ListNames", .{}, .{});
    defer result.deinit();

    var players = std.ArrayList([]const u8){};
    defer players.deinit(allocator);

    for (result.next(dbus.Array(dbus.String))) |name| {
        if (std.mem.startsWith(u8, name, "org.mpris.MediaPlayer2.")) {
            try players.append(allocator, try allocator.dupe(u8, name));
        }
    }

    return try players.toOwnedSlice(allocator);
}

File: npm-package/core/mpris.ts
import call from "./call";

export interface Metadata {
    trackid?: string;
    length?: number;
    artUrl?: string;
    album?: string;
    albumArtist?: string[];
    artist?: string[];
    asText?: string;
    audioBPM?: number;
    autoRating?: number;
    comment?: string[];
    composer?: string[];
    contentCreated?: string;
    discNumber?: number;
    firstUsed?: string;
    genre?: string[];
    lastUsed?: string;
    lyricist?: string[];
    title?: string;
    trackNumber?: number;
    url?: string;
    useCount?: number;
    userRating?: number;
}

export interface PlayerInfo {
    busName: string;
    identity?: string;
    desktopEntry?: string;
    canQuit: boolean;
    canRaise: boolean;
    canSetFullscreen: boolean;
    hasTrackList: boolean;
    fullscreen: boolean;
}

export interface PlayerStatus {
    canControl: boolean;
    canGoNext: boolean;
    canGoPrevious: boolean;
    canPause: boolean;
    canPlay: boolean;
    canSeek: boolean;
    loopStatus: "None" | "Track" | "Playlist";
    maximumRate: number;
    metadata?: Metadata;
    minimumRate: number;
    playbackStatus: "Playing" | "Paused" | "Stopped";
    position: number;
    rate: number;
    shuffle: boolean;
    volume: number;
}

/**
 * MPRIS Media Player class
 * Represents a single media player with capability checking
 */
export class Player {
    public readonly busName: string;
    public identity?: string;
    public desktopEntry?: string;

    // Capabilities
    public canQuit: boolean = false;
    public canRaise: boolean = false;
    public canSetFullscreen: boolean = false;
    public hasTrackList: boolean = false;
    public canControl: boolean = false;
    public canGoNext: boolean = false;
    public canGoPrevious: boolean = false;
    public canPlay: boolean = false;
    public canPause: boolean = false;
    public canSeek: boolean = false;

    private constructor(busName: string) {
        this.busName = busName;
    }

    /**
     * Create and initialize a Player instance
     * Fetches player capabilities on initialization
     */
    static async create(busName: string): Promise<Player> {
        const player = new Player(busName);
        await player.fetchCapabilities();
        return player;
    }

    /**
     * Fetch player capabilities and info
     */
    private async fetchCapabilities(): Promise<void> {
        try {
            const info: PlayerInfo = await call("mpris.getPlayerInfo", this.busName);
            this.identity = info.identity;
            this.desktopEntry = info.desktopEntry;
            this.canQuit = info.canQuit;
            this.canRaise = info.canRaise;
            this.canSetFullscreen = info.canSetFullscreen;
            this.hasTrackList = info.hasTrackList;

            const status: PlayerStatus = await call("mpris.getPlayerStatus", this.busName);
            this.canControl = status.canControl;
            this.canGoNext = status.canGoNext;
            this.canGoPrevious = status.canGoPrevious;
            this.canPlay = status.canPlay;
            this.canPause = status.canPause;
            this.canSeek = status.canSeek;
        } catch (error) {
            console.error(`Failed to fetch capabilities for ${this.busName}:`, error);
        }
    }

    /**
     * Brings the media player to the front
     */
    async raise(): Promise<void> {
        if (!this.canRaise) {
            throw new Error(`Player ${this.busName} does not support raise`);
        }
        return call("mpris.raise", this.busName);
    }

    /**
     * Quit the media player
     */
    async quit(): Promise<void> {
        if (!this.canQuit) {
            throw new Error(`Player ${this.busName} does not support quit`);
        }
        return call("mpris.quit", this.busName);
    }

    /**
     * Skip to the next track
     */
    async next(): Promise<void> {
        if (!this.canGoNext) {
            throw new Error(`Player ${this.busName} does not support next`);
        }
        return call("mpris.next", this.busName);
    }

    /**
     * Skip to the previous track
     */
    async previous(): Promise<void> {
        if (!this.canGoPrevious) {
            throw new Error(`Player ${this.busName} does not support previous`);
        }
        return call("mpris.previous", this.busName);
    }

    /**
     * Pause playback
     */
    async pause(): Promise<void> {
        if (!this.canPause) {
            throw new Error(`Player ${this.busName} does not support pause`);
        }
        return call("mpris.pause", this.busName);
    }

    /**
     * Toggle between play and pause
     */
    async playPause(): Promise<void> {
        if (!this.canControl) {
            throw new Error(`Player ${this.busName} cannot be controlled`);
        }
        return call("mpris.playPause", this.busName);
    }

    /**
     * Stop playback
     */
    async stop(): Promise<void> {
        if (!this.canControl) {
            throw new Error(`Player ${this.busName} cannot be controlled`);
        }
        return call("mpris.stop", this.busName);
    }

    /**
     * Start or resume playback
     */
    async play(): Promise<void> {
        if (!this.canPlay) {
            throw new Error(`Player ${this.busName} does not support play`);
        }
        return call("mpris.play", this.busName);
    }

    /**
     * Seek forward or backward
     * @param offset - The offset in microseconds
     */
    async seek(offset: number): Promise<void> {
        if (!this.canSeek) {
            throw new Error(`Player ${this.busName} does not support seek`);
        }
        return call("mpris.seek", this.busName, offset);
    }

    /**
     * Set the current track position
     * @param trackId - The track ID
     * @param position - The position in microseconds
     */
    async setPosition(trackId: string, position: number): Promise<void> {
        if (!this.canSeek) {
            throw new Error(`Player ${this.busName} does not support seek`);
        }
        return call("mpris.setPosition", this.busName, trackId, position);
    }

    /**
     * Open a URI in the media player
     */
    async openUri(uri: string): Promise<void> {
        if (!this.canControl) {
            throw new Error(`Player ${this.busName} cannot be controlled`);
        }
        return call("mpris.openUri", this.busName, uri);
    }

    /**
     * Get fullscreen status
     */
    async getFullscreen(): Promise<boolean> {
        return call("mpris.getFullscreen", this.busName);
    }

    /**
     * Set fullscreen status
     */
    async setFullscreen(value: boolean): Promise<void> {
        if (!this.canSetFullscreen) {
            throw new Error(`Player ${this.busName} does not support fullscreen`);
        }
        return call("mpris.setFullscreen", this.busName, value);
    }

    /**
     * Get playback status
     */
    async getPlaybackStatus(): Promise<"Playing" | "Paused" | "Stopped"> {
        return call("mpris.getPlaybackStatus", this.busName);
    }

    /**
     * Get loop status
     */
    async getLoopStatus(): Promise<"None" | "Track" | "Playlist"> {
        return call("mpris.getLoopStatus", this.busName);
    }

    /**
     * Set loop status
     */
    async setLoopStatus(status: "None" | "Track" | "Playlist"): Promise<void> {
        if (!this.canControl) {
            throw new Error(`Player ${this.busName} cannot be controlled`);
        }
        return call("mpris.setLoopStatus", this.busName, status);
    }

    /**
     * Get playback rate
     */
    async getRate(): Promise<number> {
        return call("mpris.getRate", this.busName);
    }

    /**
     * Set playback rate
     */
    async setRate(rate: number): Promise<void> {
        if (!this.canControl) {
            throw new Error(`Player ${this.busName} cannot be controlled`);
        }
        return call("mpris.setRate", this.busName, rate);
    }

    /**
     * Get shuffle status
     */
    async getShuffle(): Promise<boolean> {
        return call("mpris.getShuffle", this.busName);
    }

    /**
     * Set shuffle status
     */
    async setShuffle(shuffle: boolean): Promise<void> {
        if (!this.canControl) {
            throw new Error(`Player ${this.busName} cannot be controlled`);
        }
        return call("mpris.setShuffle", this.busName, shuffle);
    }

    /**
     * Get volume (0.0 to 1.0)
     */
    async getVolume(): Promise<number> {
        if (!this.canControl) {
            throw new Error(`Player ${this.busName} cannot be controlled`);
        }
        return call("mpris.getVolume", this.busName);
    }

    /**
     * Set volume (0.0 to 1.0)
     */
    async setVolume(volume: number): Promise<void> {
        if (!this.canControl) {
            throw new Error(`Player ${this.busName} cannot be controlled`);
        }
        return call("mpris.setVolume", this.busName, volume);
    }

    /**
     * Get current position in microseconds
     */
    async getPosition(): Promise<number> {
        return call("mpris.getPosition", this.busName);
    }

    /**
     * Get minimum playback rate
     */
    async getMinimumRate(): Promise<number> {
        return call("mpris.getMinimumRate", this.busName);
    }

    /**
     * Get maximum playback rate
     */
    async getMaximumRate(): Promise<number> {
        return call("mpris.getMaximumRate", this.busName);
    }

    /**
     * Get the metadata of the current track
     */
    async getMetadata(): Promise<Metadata> {
        return call("mpris.getMetadata", this.busName);
    }

    /**
     * Get player information
     */
    async getPlayerInfo(): Promise<PlayerInfo> {
        return call("mpris.getPlayerInfo", this.busName);
    }

    /**
     * Get player status including playback state and metadata
     */
    async getPlayerStatus(): Promise<PlayerStatus> {
        return call("mpris.getPlayerStatus", this.busName);
    }

    /**
     * Refresh player capabilities
     * Useful if player state might have changed
     */
    async refresh(): Promise<void> {
        await this.fetchCapabilities();
    }
}

/**
 * List all available MPRIS media players as Player instances
 */
export async function list(): Promise<Player[]> {
    const busNames: string[] = await call("mpris.listPlayers");
    const players = await Promise.all(busNames.map((busName) => Player.create(busName)));
    return players;
}
